src/assisted_test_infra/test_infra/controllers/node_controllers/oci_api_controller.py

```diff
@@ -0,0 +1,471 @@
+    4. Create a stack , declare the provision configuration.
+    5. Upload infrastructure zip file to the stack and set terraform variable
+    6. Start running job creating oci nodes discovered by redhat cluster and return terraform output files
+    7. Once nodes are discovered we configure manifest return in #6
+    """
+
+    _config: BaseOciConfig
+    # cleanup list , runs on teardown in LIFO cleanup
+    _cleanup_resources = []
+
+    def __init__(self, config: BaseNodesConfig, cluster_config: BaseClusterConfig):
+        super().__init__(config, cluster_config)
+        self.cloud_provider = None
+        self.oci_compartment_oicd = self._config.oci_compartment_oicd
+        self._initialize_oci_clients()
+
+    def _initialize_oci_clients(self):
+        """Initialize oci clients.
+
+        Added compute , volume for future usage
+        """
+        try:
+            oci_config = self._config.get_provider_config()
+            oci.config.validate_config(oci_config)
+            log.info("Initialize oci clients")
+            self.object_storage_client = oci.object_storage.ObjectStorageClient(self._config.get_provider_config())
+            self.compute_client = oci.core.ComputeClient(self._config.get_provider_config())
+            self.volume_client = oci.core.BlockstorageClient(self._config.get_provider_config())
+            # resource manager for stack creation and job
+            self.resource_manager_client = oci.resource_manager.ResourceManagerClient(
+                self._config.get_provider_config())
+            # wrapper for resource_manager_client with waiters
+            self.resource_manager_client_composite_operations = (
+                oci.resource_manager.ResourceManagerClientCompositeOperations(self.resource_manager_client))
+        except Exception as e:
+            raise e
+
+    def _create_bucket(self, name: str, public_access_type: str = "NoPublicAccess", storage_tier: str = "Standard",
+                       object_events_enabled: bool = False, versioning: str = "Disabled",
+                       auto_tiering: str = "Disabled", **kwargs):
+        bucket_details = {
+            "name": name,
+            "compartment_id": self.oci_compartment_oicd,
+            "public_access_type": public_access_type,
+            "storage_tier": storage_tier,
+            "object_events_enabled": object_events_enabled,
+            "versioning": versioning,
+            "auto_tiering": auto_tiering,
+            **kwargs
+        }
+        create_bucket_details = oci.object_storage.models.CreateBucketDetails(**bucket_details)
+        name_space = self.object_storage_client.get_namespace().data
+
+        log.info(f"Create oci bucket {create_bucket_details}")
+        obj = self.object_storage_client.create_bucket(name_space, create_bucket_details)
+        self._cleanup_resources.append(CleanupResource(self.object_storage_client.delete_bucket, name_space, name))
+        assert obj.status == 200
```

Comment:
Yes ,  the cleanup is a list of function that must run in order and run post test or on failures,  
in case create_bucket bubble up exception it wont added to cleanup resource....  but for any others case like errors 4XX , not sure about the dependency created.  
maybe its better anyway to add to the cleanup list for LIFO to keep the order , if i am going to break it the whole removal chain may break.

IMO if its not raising exception lets hanle it in the cleanup , example to try/catch the cleanup ?
What do you think ?
 
obj = self.object_storage_client.create_bucket(name_space, create_bucket_details)

Reply:
Think about case when transtaction started and aborted by the user / other cases and it "succuss" but not in the  chain.

Reply:
I added try catch to cleanup to be on the safe side